core/src/main/scala/io/projectglow/vcf/BigVCFDatasource.scala

```diff
@@ -100,4 +117,64 @@ object BigVCFDatasource extends HlsUsageLogging {
         Iterator(baos.toByteArray)
     }
   }
+
+  /**
+   * Infer sample IDs from a genomic DataFrame.
+   *
+   * - If there are no genotypes, there are no sample IDs.
+   * - If there are genotypes and sample IDs are...
+   *     - Missing, the number of missing sample IDs is the number of genotypes per row (must be the same per row).
+   *     - Present, the number of missing sample IDs is set to the number of null samples per row (must be the same).
+   *       The non-missing sample IDs are found by unifying non-null sample IDs across all rows.
+   *
+   * Missing sample IDs are represented by an empty String, as that is the default VariantContext sample ID. They will
+   * be replaced with defaults by [[VCFStreamWriter.replaceEmptySampleIds]].
+   */
+  def inferSampleIds(data: DataFrame): Seq[String] = {
+    val genotypeSchemaOpt = data
+      .schema
+      .find(_.name == "genotypes")
+      .map(_.dataType.asInstanceOf[ArrayType].elementType.asInstanceOf[StructType])
+    if (genotypeSchemaOpt.isEmpty) {
+      return Seq.empty
+    }
+
+    import data.sparkSession.implicits._
+    val hasSampleIds = genotypeSchemaOpt.get.exists(_.name == "sampleId")
+    if (hasSampleIds) {
+      val distinctSampleLists = data
+        .select("genotypes.sampleId")
+        .distinct()
+        .as[Array[String]]
+        .collect
+
+      val numMissingSampleList = distinctSampleLists.map { sl =>
```

Comment:
This behavior stills seems sort of complicated to me without a clear use case. Why would you ever have some sample ids but not others? I'm in favor of just writing the non-null sample ids.